File: 2022/puzzle9.py
from math import sqrt

def walk_the_snake(puzzle: list, rope_size: int) -> int:
    snake = [(0,0)] * rope_size
    dir_map = {'U': (0, -1), 'D': (0, 1), 'L': (-1, 0), 'R': (1, 0)}
    tail_visited = set()
    tail_visited.add(snake[0])

    for command in puzzle:
        snake[-1] = (snake[-1][0] + command[1] * dir_map[command[0]][0], snake[-1][1] + command[1] * dir_map[command[0]][1])
        anything_moved = True
        while anything_moved:
            anything_moved = False
            for i in range(len(snake) - 1, 0, -1):
                head = snake[i]
                tail = snake[i-1]
                dist_from_tail = sqrt((head[0] - tail[0]) ** 2 + (head[1] - tail[1]) ** 2)
                if dist_from_tail < 2.0: break
                # tail takes one step towards head
                tail = (tail[0] + just_one_step(head[0], tail[0]), tail[1] + just_one_step(head[1], tail[1]))
                anything_moved = True
                snake[i-1] = tail
            tail_visited.add(snake[0])
    
    return len(tail_visited)

def code_gen(data: list):
    for l in data:
        t = l.split()
        yield t[0], int(t[1])

def just_one_step(a, b) -> int:
    return 0 if a == b else int(abs(a - b)/(a - b))

with open(r"2022\puzzle9.txt") as f:
    puzzle = [command for command in code_gen(f.read().splitlines())]

print ("Part 1: {}".format(walk_the_snake(puzzle, 2)))
print ("Part 2: {}".format(walk_the_snake(puzzle, 10)))

File: 2022/y2022/src/main/java/com/chasingdings/y2022/Puzzle9.java
package com.chasingdings.y2022;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Puzzle9 extends AbstractPuzzle {
    private static final String DATA_FILE = "2022\\puzzle9.txt";
    private int part1Solution = 0;
    private int part2Solution = 0;

    @Override
    public Object solve1(String content) {
        return part1Solution;
    }

    @Override
    public Object solve2(String content) {
        return part2Solution;
    }

    @Override
    public void preprocess(String content) {
        final var puzzle = getInputDataByLine(content);
        final int ropeSize = 10;

        List<Point> snake = new ArrayList<>();
        Set<Point> part2Visited = new HashSet<>();
        Set<Point> part1Visited = new HashSet<>();
        final var origin = new Point(0, 0);

        snake = IntStream.range(0, ropeSize).mapToObj(i -> origin).collect(Collectors.toList());

        part2Visited.add(snake.get(0));
        part1Visited.add(snake.get(0));

        var curPoint = origin;

        for (var command : puzzle) {
            curPoint = movePoint(curPoint, command);

            snake.set(ropeSize - 1, curPoint);

            Boolean anythingMoved;
            do {
                anythingMoved = false;

                for (int i = ropeSize - 1; i > 0; i--) {
                    var head = snake.get(i);
                    var tail = snake.get(i - 1);

                    if (Math.sqrt(Math.pow(head.x - tail.x, 2) + Math.pow(head.y - tail.y, 2)) >= 2.0) {
                        snake.set(i - 1,
                                new Point(tail.x + justOneStep(head.x, tail.x), tail.y + justOneStep(head.y, tail.y)));
                        anythingMoved = true;
                    }
                }

                part1Visited.add(snake.get(ropeSize - 2));
                part2Visited.add(snake.get(0));
            } while (anythingMoved);

            part1Solution = part1Visited.size();
            part2Solution = part2Visited.size();
        }
    }

    private Point movePoint(Point c, String cmd) {
        char d = cmd.charAt(0);
        int dt = Integer.parseInt(cmd.substring(2));
        return d == 'U' ? new Point(c.x, c.y - dt)
                : d == 'D' ? new Point(c.x, c.y + dt) : d == 'L' ? new Point(c.x - dt, c.y) : new Point(c.x + dt, c.y);
    }

    private int justOneStep(int a, int b) {
        return (a == b) ? 0 : a > b ? 1 : -1;
    }

    @Override
    public String getDataFilePath() {
        return DATA_FILE;
    }

    @Override
    public String getPuzzleName() {
        return "Day 9 - Rope Bridge";
    }

    final class Point {
        int x;
        int y;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        // add equals
        public boolean equals(Object o) {
            if (o == this)
                return true;
            if (!(o instanceof Point)) {
                return false;
            }

            Point point = (Point) o;

            return point.x == x && point.y == y;
        }

        // add hashCode
        public int hashCode() {
            int result = 17;
            result = 31 * result + x;
            result = 31 * result + y;
            return result;
        }
    }
}
